File: C++/섹션09. 2022 하반기 이후 (A형 풀이)/60. 택배 하차 (코드트리, 2025 하반기 오전 1번).cpp
#include <stdio.h>

#define MAX (50 + 5)

int T;
int N, M, maxBoxNum;

int MAP[MAX][MAX];

struct BOX
{
	int k; // 택배 번호
	int h; // 세로 크기
	int w; // 가로 크기
	int r; // 행
	int c; // 열
	bool drop; // 하차 여부
};

BOX box[100 + 10];

void input()
{
	maxBoxNum = 0;

	scanf("%d %d", &N, &M);

	for (int i = 1; i <= 100; i++)
		box[i].k = 0;

	for (int r = 0; r <= N + 1; r++)
		for (int c = 0; c <= N + 1; c++)
			MAP[r][c] = -1;

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			MAP[r][c] = 0;
}

void printMap() // for debug
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

bool checkDown(int index)
{
	BOX b = box[index];

	for (int c = b.c; c < b.c + b.w; c++)
		if (MAP[b.r + b.h][c] != 0) return false; // 이동 불가

	return true;
}

void setBox(int index)
{
	BOX b = box[index];
	for (int r = b.r; r < b.r + b.h; r++)
		for (int c = b.c; c < b.c + b.w; c++)
			MAP[r][c] = index;
}

void deleteBox(int index)
{
	BOX b = box[index];
	for (int r = b.r; r < b.r + b.h; r++)
		for (int c = b.c; c < b.c + b.w; c++)
			MAP[r][c] = 0;
}

void moveDown(int index)
{
	BOX b = box[index];

	deleteBox(index);

	while (1)
	{
		if (checkDown(index) == false) break;

		b.r += 1; // 임시 변수 수정
		box[index].r += 1; // 실제값 수정
	}
	
	setBox(index);
}

bool checkLeft(int index)
{
	BOX b = box[index];

	for (int r = b.r; r < b.r + b.h; r++)
		for (int c = 1; c < b.c; c++)
			if (MAP[r][c] != 0) return false;

	return true;
}

void moveLeft()
{
	for (int i = 1; i <= maxBoxNum; i++)
	{
		if (box[i].drop == true || box[i].k == 0) continue;

		if (checkLeft(i) == true)
		{
			box[i].drop = true;
			deleteBox(i);

			printf("%d\n", i);

			return;
		}
	}
}

bool checkRight(int index)
{
	BOX b = box[index];

	for (int r = b.r; r < b.r + b.h; r++)
		for (int c = b.c + b.w; c <= N; c++)
			if (MAP[r][c] != 0) return false;

	return true;
}

void moveRight()
{
	for (int i = 1; i <= maxBoxNum; i++)
	{
		if (box[i].drop == true || box[i].k == 0) continue;

		if (checkRight(i) == true)
		{
			box[i].drop = true;
			deleteBox(i);

			printf("%d\n", i);

			return;
		}
	}
}

void moveDownAll()
{
	while (1)
	{
		bool check = true;
		for (int i = 1; i <= maxBoxNum; i++)
		{
			if (box[i].drop == true || box[i].k == 0) continue;
			if (checkDown(i) == false) continue;

			check = false;

			moveDown(i);
		}

		if (check == true) break;
	}
}

void simulate()
{
	// 1. 택배 투입
	for (int m = 0; m < M; m++)
	{
		int k, h, w, c;

		scanf("%d %d %d %d", &k, &h, &w, &c);

		box[k].k = k; // M과 최대 k가 다를 수 있음.
		box[k].h = h;
		box[k].w = w;
		box[k].r = 1;
		box[k].c = c;
		box[k].drop = false;

		if (maxBoxNum < k) maxBoxNum = k;

		moveDown(k);
	}

	// printMap();

	for (int m = 0; m < M; m += 2)
	{
		// 2. 택배 하차 (좌측)
		moveLeft();
		moveDownAll();

		// 3. 택배 하차 (우측)
		moveRight();
		moveDownAll();
	}
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		simulate();
	}

	return 0;
}

File: Python/섹션09. 2022 하반기 이후 (A형 풀이)/60. 택배 하차 (코드트리, 2025 하반기 오전 1번).py
MAX = 50 + 5

MAP = [[0] * MAX for _ in range(MAX)] 

class BOX:
    def __init__(self, k: int, h: int, w: int, r: int, c: int, drop: bool):
        self.k = k
        self.h = h
        self.w = w
        self.r = r
        self.c = c
        self.drop = drop

box = [BOX(0, 0, 0, 0, 0, False) for _ in range(100 + 10)]

def input_data():
        global N, M, max_box_num, MAP, box
        max_box_num = 0

        N, M = map(int, input().split())
                
        for i in range(1, 101):
            box[i].k = 0

        for r in range(0, N + 2):
            for c in range(0, N + 2):
                MAP[r][c] = -1

        for r in range(1, N + 1):
            for c in range(1, N + 1):
                MAP[r][c] = 0
                
def print_map(): # for debug
    for r in range(1, N + 1):
        print(*MAP[r][1:N + 1])
    print()                
                
def check_down(index):
    b = box[index]
    for c in range(b.c, b.c + b.w):
        if MAP[b.r + b.h][c] != 0:
            return False
    return True

def set_box(index):
    b = box[index]
    for r in range(b.r, b.r + b.h):
        for c in range(b.c, b.c + b.w):
            MAP[r][c] = index

def delete_box(index):
    b = box[index]
    for r in range(b.r, b.r + b.h):
        for c in range(b.c, b.c + b.w):
            MAP[r][c] = 0

def move_down(index):
    b = box[index]
    delete_box(index)

    while True:
        if not check_down(index):
            break
        b.r += 1
        
    set_box(index)

def check_left(index):
    b = box[index]
    for r in range(b.r, b.r + b.h):
        for c in range(1, b.c):
            if MAP[r][c] != 0:
                return False
    return True

def move_left():
    global max_box_num
    for i in range(1, max_box_num + 1):
        if box[i].drop or box[i].k == 0:
            continue
        if check_left(i):
            box[i].drop = True
            delete_box(i)
            print(i)
            return

def check_right(index):
    b = box[index]
    for r in range(b.r, b.r + b.h):
        for c in range(b.c + b.w, N + 1):
            if MAP[r][c] != 0:
                return False
    return True

def move_right():
    global max_box_num
    for i in range(1, max_box_num + 1):
        if box[i].drop or box[i].k == 0:
            continue
        if check_right(i):
            box[i].drop = True
            delete_box(i)
            print(i)
            return

def move_down_all():
    while True:
        check = True
        for i in range(1, max_box_num + 1):
            if box[i].drop or box[i].k == 0:
                continue
            if not check_down(i):
                continue
            check = False
            move_down(i)
        if check:
            break

def simulate():
    global max_box_num

    # 1. 택배 투입
    for _ in range(M):
        k, h, w, c = map(int, input().split())

        box[k].k = k # M과 최대 k가 다를 수 있음.
        box[k].h = h
        box[k].w = w
        box[k].r = 1
        box[k].c = c
        box[k].drop = False

        if max_box_num < k: max_box_num = k

        move_down(k)

    for _ in range(0, M, 2):
        # 2. 택배 하차 (좌측)
        move_left()
        move_down_all()

        # 3. 택배 하차 (우측)
        move_right()
        move_down_all()
                                
# T = int(input())
T = 1
for tc in range(T):  
    input_data()
        
    simulate()